src/backend/commands/explain.c

```diff
@@ -1218,39 +1218,10 @@ ExplainNode(PlanState *planstate, List *ancestors,
 		 * Estimates will have to be scaled down to be per-segment (except in a
 		 * few cases).
 		 */
-		if ((plan->directDispatch).isDirectDispatch)
+		if (es->currentSlice)
 		{
-			scaleFactor = 1.0;
-		}
-		else if (plan->flow != NULL && CdbPathLocus_IsBottleneck(*(plan->flow)))
-		{
-			/*
-			 * Data is unified in one place (singleQE or QD), or executed on a
-			 * single segment.  We scale up estimates to make it global.  We
-			 * will later amend this for Motion nodes.
-			 */
-			scaleFactor = 1.0;
-		}
-		else if (plan->flow != NULL && CdbPathLocus_IsSegmentGeneral(*(plan->flow)))
-		{
-			/* Replicated table has full data on every segment */
-			scaleFactor = 1.0;
-		}
-		else if (plan->flow != NULL && es->pstmt->planGen == PLANGEN_PLANNER)
-		{
-			/*
-			 * The plan node is executed on multiple nodes, so scale down the
-			 * number of rows seen by each segment
-			 */
-			scaleFactor = CdbPathLocus_NumSegments(*(plan->flow));
-		}
-		else
-		{
-			/*
-			 * The plan node is executed on multiple nodes, so scale down the
-			 * number of rows seen by each segment
-			 */
-			scaleFactor = getgpsegmentCount();
+			if (es->currentSlice->gangType != GANGTYPE_UNALLOCATED)
```

Comment:
- Can you tell more about why the logic around `es->currentSlice` can replace all old conditions, especially default one?
- If it really works, the comment above should be somehow reworked and explain it - the old comment points to old logic.
- If it really works, all 3 `if` statements can be squashed to one.

Reply:
We divide the number of rows when a plan node is executed on multiple segments. Here we simply look at the number of segments on which a given plan node runs and set the `scaleFactor` to it.  The previous logic does the same thing. It checks, using `flow`, whether the node is running on a single segment or on several segments. If several, it takes the number of segments as the `scaleFactor`. The only trouble is that flow is missing in some nodes of the plan generated by orca. 

The new logic does the following: Checks if we are in any slice and if we are not in QD. Then our node runs on several segments and we take the `scaleFactor` for the number of these segments.

In the old logic, it checks the flow. If it equals singleton, then the node is executed on one segment, otherwise (if the data is not replicated) on several segments. As far as I understand, if data is replicated, then there is no logic to process it in parallel in one slice on several segments, so the logic is correct.

Reply:
> if data is replicated, then there is no logic to process it in parallel in one slice on several segments

We can easily implement this behavior when joining a replicated table with the distributed one:
```sql
create table tr(a int) distributed replicated;                       
insert into tr select generate_series(1, 1000000);
create table ts(a int) distributed by (a);                                                       
insert into ts select generate_series(1, 20);

set optimizer = on;
explain analyze select * from ts join tr using(a);
                                                          QUERY PLAN                                                          
------------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..957.27 rows=20 width=4) (actual time=765.845..778.837 rows=20 loops=1)
   ->  Hash Join  (cost=0.00..957.27 rows=7 width=4) (actual time=1.048..763.572 rows=8 loops=1)
         Hash Cond: (tr.a = ts.a)
         Extra Text: (seg0)   Hash chain length 1.0 avg, 1 max, using 8 of 524288 buckets.
         ->  Result  (cost=0.00..467.53 rows=333334 width=4) (actual time=0.015..594.190 rows=334042 loops=1)
               ->  Seq Scan on tr  (cost=0.00..449.70 rows=1000000 width=4) (actual time=0.012..332.762 rows=1000000 loops=1)
         ->  Hash  (cost=431.00..431.00 rows=7 width=4) (actual time=0.020..0.020 rows=8 loops=1)
               ->  Seq Scan on ts  (cost=0.00..431.00 rows=7 width=4) (actual time=0.008..0.009 rows=8 loops=1)
 Planning time: 38.300 ms
   (slice0)    Executor memory: 59K bytes.
   (slice1)    Executor memory: 4170K bytes avg x 3 workers, 4170K bytes max (seg0).  Work_mem: 1K bytes max.
 Memory used:  128000kB
 Optimizer: Pivotal Optimizer (GPORCA)
 Execution time: 779.678 ms
(14 rows)
```
This plan was generated on the actual commit https://github.com/arenadata/gpdb/commit/e32297eb4d2f48882efb9bacec0a6b79829e2759 of the current branch. As we can see ORCA uses an incorrect scale factor for replicated tables: get 1kk/3 rows instead of 1kk (fixed in https://github.com/greenplum-db/gpdb/commit/5bd7930dede0106ab0833e27048ee31cdb3c37fd).

BTW, Postgres planner returns the correct number of rows:
```sql
set optimizer = off;
explain analyze select * from ts join tr using(a);
                                                            QUERY PLAN                                                            
----------------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=3.45..14854.65 rows=20 width=4) (actual time=1207.009..1235.894 rows=20 loops=1)
   ->  Hash Join  (cost=3.45..14854.65 rows=7 width=4) (actual time=0.665..1205.313 rows=8 loops=1)
         Hash Cond: (tr.a = ts.a)
         Extra Text: (seg0)   Hash chain length 1.0 avg, 1 max, using 8 of 524288 buckets.
         ->  Seq Scan on tr  (cost=0.00..11101.00 rows=1000000 width=4) (actual time=0.034..424.788 rows=1000000 loops=1)
         ->  Hash  (cost=3.20..3.20 rows=7 width=4) (actual time=0.043..0.043 rows=8 loops=1)
               ->  Seq Scan on ts  (cost=0.00..3.20 rows=7 width=4) (actual time=0.020..0.025 rows=8 loops=1)
 Planning time: 6.201 ms
   (slice0)    Executor memory: 59K bytes.
   (slice1)    Executor memory: 4170K bytes avg x 3 workers, 4170K bytes max (seg0).  Work_mem: 1K bytes max.
 Memory used:  128000kB
 Optimizer: Postgres query optimizer
 Execution time: 1237.148 ms
(13 rows)
```

Honestly, I think that your solution is incorrect and we can't simply remove these complicated scale factor checks, as you have suggested.